### DISCUSSION TITLE
How to calculate buffer region?

### DISCUSSION THREAD

**Discussion by mdkumargce** [roleoroleo/yi-hack-MStar#363] [Q&A] [answered]
Hi roleoroleo,

I'm working on compiling a custom firmware for Yi-Iot camera(Victure - SC220) based on realtek MIPS architecture. 
I managed to get a ssh connection through dropbear.
Can you please tell me how to calculate the offsets and sizes of a buffer? it really helps if you can help me figure out the values for the variables listed below.
BUF_OFFSET
BUF_SIZE
FRAME_HEADER_SIZE
DATA_OFFSET
LOWRES_BYTE
HIGHRES_BYTE

Thanks,
Dinesh

**Comment by roleoroleo** [comment]
Where do you get these parameters in the MStar firmware?

**Reply by mdkumargce** [reply]
I'm sorry I got this from Allwinner firmware and posted here accidentally.

I have a circular buffer file "fshare_frame_buf" under /dev/shm and looking it in hexeditor I can guess the buf_offset being 300 but I couldn't figure out rest of the parameters. your help is highly appreciated.

**Comment by roleoroleo** [comment, marked as answer]
BUF_SIZE is the size of the file.

Then you have to detect a frame and check the header.
- Search for the sequence 00 00 00 01 67 (NALU SPS).
- After this NALU you should see a PPS 00 00 00 01 68 and probably an IDR 00 00 00 01 65.
- Normally (with these cams) a PPS is 8 bytes long (eg. 00 00 00 01 68 EE 3C 80 E1).
- The distance between the end of the PPS and the start of the IDR (or another NALU that follows PPS) is an header.

FRAME_HEADER_SIZE is the size of this header.
LOWRES_BYTE is the value of the byte for low res frames (normally 8).
HIGHRES_BYTE is the value of the byte for high res frames (normally 4).
DATA_OFFSET is the offset of the resolution byte inside the header from the beginning of the header - 17 (start from 0).

Let me explain with an example:
08 00 00 00 B5 95 E1 04 50 BE 92 5F FA 68 37 55 04 **04** 41 2B 00 00

08 00 00 00 - The first 4 bytes are the length of the frame.
The highlighted number is the resolution byte: 04 means high res frame.
41 2B is the frame counter (there are 2 different counter, 1 for high res and 1 for low res).
DATA_OFFSET is 0 because if you count the bytes (starting from 0), the resolution byte is the 17th.

**Reply by mdkumargce** [reply]
Thank you very much.
